libraries/chain/webassembly/runtimes/eos-vm-oc/ipc_helpers.cpp

```diff
@@ -89,7 +89,7 @@ bool write_message_with_fds(int fd_to_send_to, const eosvmoc_message& message, c
       .iov_len = sz
    };
    union {
-      char buf[CMSG_SPACE(max_num_fds * sizeof(int))];
+      char buf[CMSG_SPACE(max_num_fds * sizeof(int))] = {};
```

Comment:
This was uninitialized when called from ipc_helpers.cpp:115.

In line 115 `wrote = sendmsg(fd_to_send_to, &msg, 0);` msg is used.
`msg` has member: `msg_control` which is initialized in line 99: `msg.msg_control = u.buf;`.
However memeber of union `u.buf` is unitialized (declaration in line 94).

This happens in every scenario where function `write_message_with_fds` is being used.

Reply:
I am not seeing why uninitialized memory was accessed? At line 104, we set `cmsg->cmsg_len = CMSG_LEN(sizeof(int) * fds.size());`, which is the same size we initialize in the loop at line 106. So I assume that the `sendmsg` function should only access that area of `u.buf` that has been initialized. Zeroing the whole buffer might be overkill.

Reply:
Right. My guess is valgrind just isn't smart enough to fully reason on the behavior of this syscall. But it's probably best practice to initialize it anyways. It's small and not performance sensitive.

Reply:
I confirm (sizeof(int) * fds.size()) == 8 in my case.
Also from debugger: CMSG_LEN(sizeof(int) * fds.size()) == 24.
Apparently last 8 bytes of u.msg are not initialized and they are being looked at by sendmsg.
I don't have sendmsg code in debugger, but probably alternative / better way would be to set msg.msg_controllen to 24 from 32 in our case? However the logic would be harder to read i.e. something like this after end of the loop
` msg.msg_controllen = (void*) p - (void*) cmsg;`

Details from debugger, in my case:
u.buf -> points to i.e. 0x1ffeffa4b0 - 32 bytes
cmsg -> 0x1ffeffa4b0 - 32 bytes
cmsg->cmsg_len -> 0x1ffeffa4b0 - initialized 8 bytes
cmsg->cmsg_level -> 0x1ffeffa4b8 - initialized 4 bytes
cmsg->cmsg_type -> 0x1ffeffa4bc - initialized 4 bytes
CMSG_DATA(cmsg) -> 0x1ffeffa4c0 - initialized 8 bytes total in the loop = 2x sizeof(thisfd)

Reply:
I couldn't find where `CMSG_LEN` is defined, but `CMSG_LEN(sizeof(int) * fds.size()) == 24` seems wrong.
Also just noticed that `max_num_fds == 4` so yes pre-zeroing the buffer of 4 ints should be fast enough.

Reply:
CMSG_LEN definition: https://github.com/openSUSE/kernel/blob/master/include/linux/socket.h#L122
And help is here: https://linux.die.net/man/3/cmsg_space
"first initialize the msg_controllen member of the msghdr with the length of the control message buffer"
so msg.msg_controllen looks good.
The same for CMSG_LEN: "In each control message, initialize cmsg_len (with CMSG_LEN()), the other cmsghdr header fields, and the data portion using CMSG_DATA().".

So I don't see any other way than just initialize whole u.buf (or at least 8 unused bytes).

Reply:
Thanks @mikelik , I didn't realize that was a system macro. And yes, it does make sense to initialize the whole `u.buf`. Thanks for the explanation!

Reply:
Actually, after looking at the doc you referred to, I do think it would be more correct to set:

`msg.msg_controllen = p - CMSG_DATA(cmsg);`

After the loop (so that we correctly only send the descriptors we initialized, and not extra descriptors set to 0).

Reply:
The proper way is shown in the example at the bottom: https://linux.die.net/man/3/cmsg_space